### PR TITLE
Add win-ci pipeline for on-device training

### PR DESCRIPTION
**Pull request by baijumeswani**
Adding windows CI pipeline for on device training build that will build with `--enable_training_on_device` flag and run the tests.

### CODE REVIEW

File: onnxruntime/core/session/onnxruntime_c_api.cc [hunk 1]
```diff
@@ -2246,6 +2246,10 @@ ORT_API_STATUS_IMPL(OrtApis::SessionOptionsSetCustomJoinThreadFn, _Inout_ OrtSes
   API_IMPL_END
 }
 
+#ifdef _WIN32
+#pragma warning(push)
+#pragma warning(disable : 4702)
+#endif
```
Comment (askhade) [inline onnxruntime/core/session/onnxruntime_c_api.cc:2252]: adding else should work... you dont need to disable the warning

File: tools/ci_build/github/azure-pipelines/win-ci-on-device-training.yml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+parameters:
+- name: RunOnnxRuntimeTests
+  displayName: Run Tests?
+  type: boolean
+  default: true
+
+stages:
+- stage: training_x64_debug
+  dependsOn: []
```
Comment (mszhanyi) [inline tools/ci_build/github/azure-pipelines/win-ci-on-device-training.yml:9]: It looks there's only one stage, the ` dependsOn: []` is useless

File: tools/ci_build/github/azure-pipelines/win-ci-on-device-training.yml [hunk 3]
```diff
@@ -0,0 +1,21 @@
+parameters:
+- name: RunOnnxRuntimeTests
```
Comment (mszhanyi) [inline tools/ci_build/github/azure-pipelines/win-ci-on-device-training.yml:2]: I suggest it'd better to split the workflow into 2 stage, building & testing in the next PRs.
some stages could be skipped in pipeline portal and different stages could run in different pools.

Comment (baijumeswani) [reply tools/ci_build/github/azure-pipelines/win-ci-on-device-training.yml:2]: I didn't quite follow the suggestion.
This file is just a modification of an existing file `tools/ci_build/github/azure-pipelines/win-ci-pipeline.yml` (Windows CPU CI Pipeline for ort) for the on-device-training scenario (addition of the `--enable_training_on_device` flag in the build args). As such, if you look at the template it invokes: `templates/win-cpu-ci.yml`, the build happens in a different step compared to the run tests step.
You can view the pipeline run [here](https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=717535&view=logs&j=78245837-8ee8-5268-5ba6-4858d2b4c8d6&t=172d9be8-96dc-5dc5-4044-7f56c5e4e79c).

Maybe I missed what you are trying to suggest.

Comment (snnn) [reply tools/ci_build/github/azure-pipelines/win-ci-on-device-training.yml:2]: win-cpu-ci.yml is not updated to support it yet. I think the concern can be address in a new PR, for all Windows GPU build pipelines.

**Review by snnn — DISMISSED** [review]

File: orttraining/orttraining/training_api/checkpoint.cc [hunk 4]
```diff
@@ -182,11 +182,11 @@ Status OrtSaveInternal(
     const std::vector<ONNX_NAMESPACE::TensorProto>& non_trainable_tensor_protos,
     const PathString& checkpoint_path) {
   // Make sure name unique across trainable and non-trainable lists.
-  std::unordered_set<std::string> trainable_unique_names;
-  std::unordered_set<std::string> non_trainable_unique_names;
+  std::set<std::string> trainable_unique_names;
```
Comment (askhade) [inline orttraining/orttraining/training_api/checkpoint.cc:185]: curious why is this change required?

File: tools/ci_build/github/azure-pipelines/win-ci-pipeline.yml [hunk 5]
```diff
@@ -130,3 +130,18 @@ stages:
...
+
+- stage: training_x64_debug_on_device_training
+  dependsOn: []
+  jobs:
+    - template: templates/win-cpu-ci.yml
+      parameters:
+        BuildConfig: 'Debug'
+        EnvSetupScript: setup_env.bat
+        buildArch: x64
+        additionalBuildFlags: --enable_training --enable_training_on_device --build_wheel
```
Comment (askhade) [inline tools/ci_build/github/azure-pipelines/win-ci-pipeline.yml:142]: do we need build_wheel here?

File: orttraining/orttraining/training_api/checkpoint.cc (hunk 4 above)
Comment (baijumeswani) [reply orttraining/orttraining/training_api/checkpoint.cc:185]: The problem is that we perform `set_intersection` on the `unordered_set`. The definition from https://en.cppreference.com/w/cpp/algorithm/set_intersection

> Constructs a sorted range beginning at d_first consisting of elements that are found in both sorted ranges [first1, last1) and [first2, last2).

Is expecting the inputs to be sorted. While the linux build succeeds, the windows build throws an exception at runtime:

> Assertion failed: sequence not ordered

More details [here](https://stackoverflow.com/questions/48158811/c-library-method-for-intersection-of-two-unordered-set/48158882#48158882).

To circumvent this, we can either loop over the `unordered_set` manually, or use `set` instead. I chose the latter since this is a one-time expense to find the intersection.

File: tools/ci_build/github/azure-pipelines/win-ci-pipeline.yml (hunk 5 above)
Comment (baijumeswani) [reply tools/ci_build/github/azure-pipelines/win-ci-pipeline.yml:142]: I don't think we need. I kept it to be consistent with the other stages in this file.
Can remove.

File: tools/ci_build/github/azure-pipelines/win-ci-on-device-training.yml (hunk 3 above)
Comment (snnn) [reply tools/ci_build/github/azure-pipelines/win-ci-on-device-training.yml]: @baijumeswani, I think he means we should use cpu machines for building code, gpu machine for testing, given the number of GPU machines is limited.

File: orttraining/orttraining/test/training_api/trainer/trainer.cc [hunk 6]
```diff
@@ -333,8 +334,9 @@ int RunTraining(const TestRunnerParameters& params) {
 
       if ((batch_idx + 1) % params.checkpoint_interval == 0) {
         // Save trained weights
-        PathString ckpt_file = ToPathString(
-            params.output_dir + "/ckpt_" + params.model_name + std::to_string(batch_idx));
+        std::basic_ostringstream<PathChar> oss;
+        oss << ToPathString("/ckpt_") << ToPathString(params.model_name) << ToPathString(std::to_string(batch_idx));
```
Comment (pengwa) [inline orttraining/orttraining/test/training_api/trainer/trainer.cc:338]: nit: I belive the prefixing / required since we used ConcatPathComponent. Right?

File: orttraining/orttraining/test/training_api/trainer/trainer.cc [hunk 7]
```diff
@@ -356,7 +358,9 @@ int RunTraining(const TestRunnerParameters& params) {
   }
 
   // Save trained weights
-  PathString ckpt_file = ToPathString(params.output_dir + "/ckpt_" + params.model_name);
+  std::basic_ostringstream<PathChar> oss;
+  oss << ToPathString("/ckpt_") << ToPathString(params.model_name);
```
Comment (pengwa) [inline orttraining/orttraining/test/training_api/trainer/trainer.cc:362]: nit: same here.

File: orttraining/orttraining/test/training_api/trainer/trainer.cc (hunk 6 above)
Comment (pengwa) [reply orttraining/orttraining/test/training_api/trainer/trainer.cc:338]: nit: And maybe we could use std::ostringstream for oss, afterward, we can use one single ToPathString on it.

File: tools/ci_build/github/azure-pipelines/win-ci-pipeline.yml [hunk 8]
```diff
@@ -130,3 +130,18 @@ stages:
         RunOnnxRuntimeTests: ${{ parameters.RunOnnxRuntimeTests }}
         # Liqun is fixing them
         RunStaticCodeAnalysis: true
+
+- stage: training_x64_debug_on_device_training
+  dependsOn: []
+  jobs:
+    - template: templates/win-cpu-ci.yml
+      parameters:
+        BuildConfig: 'Debug'
```
Comment (pengwa) [inline tools/ci_build/github/azure-pipelines/win-ci-pipeline.yml:139]: So we prefer Debug build validation against Release?

Comment (baijumeswani) [reply tools/ci_build/github/azure-pipelines/win-ci-pipeline.yml:139]: I just picked one of the two. I don't think we need to test both in the pipeline (at least for now). Can change to release.

Comment (askhade) [reply tools/ci_build/github/azure-pipelines/win-ci-pipeline.yml]: Debug is good for the purpose of this pipeline... we will add another packaging pipeline in future which will use release build.

**Review by askhade — APPROVED** [review]

File: tools/ci_build/github/azure-pipelines/win-ci-on-device-training.yml (hunk 3 above)
Comment (baijumeswani) [reply tools/ci_build/github/azure-pipelines/win-ci-on-device-training.yml]: Thanks. may we pick up this work in a follow up PR?

**Review by snnn — APPROVED** [review]

**Comment by baijumeswani** [comment]
Thank you for the review. :)
